ray_on_golem/server/services/golem/golem.py

```diff
@@ -311,14 +319,77 @@ async def _upload_node_configuration(
+        provider_desc = await self.get_provider_desc(context.activity)
         logger.info("Starting ssh service on " f"{provider_desc}, {ip=}, {context.activity=}")
-        await add_state_log("[7/9] Starting ssh service...")
         await context.run("service ssh start")
 
+    async def _restart_ssh_server(self, context: WorkContext, ip: str):
+        provider_desc = await self.get_provider_desc(context.activity)
+        logger.debug(f"Restarting ssh service on {provider_desc}, {ip=}, {context.activity=}")
+        try:
+            await context.run("service ssh restart", timeout=120)
+            logger.debug(
+                f"Restarting ssh service on {provider_desc}, {ip=}, {context.activity=} done"
+            )
+        except Exception:
+            msg = f"Failed to restart SSH server {provider_desc}, {ip=}, {context.activity=}"
+            logger.warning(msg)
+            logger.debug(msg, exc_info=True)
+
+    @staticmethod
+    async def _verify_ssh_connection_check(activity_id, ssh_command):
+        process = await asyncio.create_subprocess_shell(
+            ssh_command,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+        )
+
+        stdout, stderr = await process.communicate()
+
+        debug_data = f"{activity_id=}, exitcode={process.returncode}, {stdout=}, {stderr=}"
+        logger.debug(debug_data)
+
+        if process.returncode != 0:
+            raise Exception(f"SSH connection check failed. {debug_data}")
+
+    async def _sentry_ssh_connection(
+        self,
+        context: WorkContext,
+        ip: str,
+        ssh_proxy_command: str,
+        ssh_user: str,
+        ssh_private_key_path: Path,
+    ):
+        provider_desc = await self.get_provider_desc(context.activity)
+        ssh_command = (
+            f"{get_ssh_command(ip, ssh_proxy_command, ssh_user, ssh_private_key_path)} uptime"
+        )
+
+        fails_count = 0
+        while True:
+            try:
+                await self._verify_ssh_connection_check(context.activity.id, ssh_command)
+            except Exception:
+                fails_count += 1
+                if fails_count >= 5:
+                    # Ray's autoscaler won't notice but we will at least save some GLMs
+                    msg = f"Destroying activity due to no SSH connection to {provider_desc}"
+                    logger.warning(msg)
+                    logger.debug(msg, exc_info=True)
+                    try:
+                        await context.activity.destroy()
+                    except Exception:
+                        logger.debug(f"Cannot destroy activity {provider_desc}", exc_info=True)
+                    break
+
+                logger.debug(
+                    f"SSH connection to {provider_desc} stopped working. Restarting SSH server",
+                    exc_info=True,
+                )
+                await self._restart_ssh_server(context, ip)
```

Comment:
Why are we restarting the ssh server after the first failure?

It should at least try a few times to connect to rule out temporary connection issues on the network. Then, after deciding that ssh server looks like it's not working, try to restart it, and do the whole trip a few times.

Like utilizing `_verify_ssh_connection` under the hood.

Reply:
I switch `_verify_ssh_connection_check` to `_verify_ssh_connection`